src/lfortran/parser/semantics.h

```diff
@@ -63,6 +63,27 @@ static inline char* name2char(const ast_t *n)
     return down_cast2<Name_t>(n)->m_id;
 }
 
+static inline void check_names(const char* n1, const char* n2, Location &loc)
+{       
+        if (n2) {
+                if (strcmp(n1, n2) != 0) {
+                        throw LCompilers::LFortran::parser_local::ParserError(
+                        "End subroutine name does not match subroutine name", loc);
+                }
+        }
+}
+
+static inline char* name2char_check(const ast_t *n1, const ast_t *n2, Location &loc)
+{
+        const char* n1c = down_cast2<Name_t>(n1)->m_id;
+        const char* n2c = down_cast2<Name_t>(n1)->m_id;
+        if(n2) {
+                n2c = down_cast2<Name_t>(n2)->m_id;
+        } 
+        check_names(n1c, n2c, loc);
+        return down_cast2<Name_t>(n1)->m_id;
+}
```

Comment:
```suggestion
static inline void check_names(char* n1, char* n2, Location &loc) {
    if (strcmp(n1, n2) != 0) {
        throw LCompilers::LFortran::parser_local::ParserError(
            "End subroutine name does not match subroutine name", loc);
    }
}

static inline char* name2char_check(const ast_t *n1, const ast_t *n2,
        Location &loc) {
    char* n1c = down_cast2<Name_t>(n1)->m_id;
    if(n2) {
        char* n2c = down_cast2<Name_t>(n2)->m_id;
        check_names(n1c, n2c, loc);
    }
    return n1c;
}
```
We use 4 spaces for tab, can you please update it in your editor.